lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1308,10 +1308,10 @@ static bool performCompile(CompilerInstance &Instance,
 
   // FIXME: This is still a lousy approximation of whether the module file will
   // be externally consumed.
-  bool moduleIsPublic =
-      !Instance.getMainModule()->hasEntryPoint() &&
+  bool moduleIsPublic = opts.ObjcCompatibilityHeaderIsPublic ||
+      (!Instance.getMainModule()->hasEntryPoint() &&
       opts.ImplicitObjCHeaderPath.empty() &&
-      !Context.LangOpts.EnableAppExtensionRestrictions;
+      !Context.LangOpts.EnableAppExtensionRestrictions);
```

Comment:
Should this be a tri-state option (public only, public and internal, infer from other flags)? That way we could start getting build systems to tell us what they want and perhaps eventually deprecate or remove the "lousy approximation".

(Also, you might want to tweak either the comment or the precise format of the code to reflect that the part after the `||` is what's lousy.)

Reply:
@brentdax Thank you for taking a look! Thinking more about this patch, it seems we should emit three headers probably all the time: API-only header, SPI-only header, and an internal header. The build system should be able to install the first two into the correct directory and always omit the third. I plan to close this PR and open another one. Will address your comments there!

Reply:
If you're going that direction, I'd create separate flags—`-emit-public-objc-header`, `-emit-spi-objc-header`, `-emit-internal-objc-header`—and allow the build system to specify whichever combination it wants to get. Then we can leave `-emit-objc-header` as a flag that means either `-emit-public-objc-header` or `-emit-internal-objc-header` based on the lousy approximation, with a deprecation warning when we're ready to do that.